connectionmgr.go

```diff
@@ -0,0 +1,72 @@
+package utils
+
+import (
+	"log"
+	"sync"
+)
+
+// Initializer provides an interface an object to provide methods for startup and shutdown.
+type Initializer interface {
+	Initialize()
+	Close()
+}
+
+// CloseFunc is a function that gracefully shuts down a connection as a side effect.
+type CloseFunc func()
+
+// ConnectionManager manages initialization and shutdown of long lived connections.
+// Each connection object must match the Initializer interface. Initialization happens
+// in parallel. The waitgroup can be used to wait util all connections are initialized.
+// Close kick off the proper shutdown of all managed connections.
+type ConnectionManager struct {
+	closerChan chan CloseFunc
+	Close      CloseFunc
+	WG         sync.WaitGroup
+}
+
+// NewConnectionManager initializes a new ConnectionManager object that can be used
+// to manage the life of long lived remote connections such as to a database.
+func NewConnectionManager(logger *log.Logger) ConnectionManager {
+	closerChan := make(chan CloseFunc)
+	shutdown := make(chan struct{})
+	var stopwg sync.WaitGroup
+	go func() {
+		closers := []CloseFunc{}
+	loop:
+		for {
+			select {
+			case <-shutdown:
+				logger.Println("Shutting Down")
+				break loop
+			case c := <-closerChan:
+				stopwg.Add(1)
+				closers = append(closers, c)
+			}
+		}
+
+		for _, c := range closers {
+			go func(c func()) {
+				c()
+				stopwg.Done()
+			}(c)
+		}
+	}()
+	return ConnectionManager{
+		closerChan: closerChan,
+		Close: func() {
+			shutdown <- struct{}{}
+			stopwg.Wait()
+		},
+	}
+}
+
+// DoInit initializes an object matching the Initializer interface, setting its close
+// operation to run when the Connection Manager's Close method is called.
+func (cm *ConnectionManager) DoInit(initializer Initializer) {
```

Comment:
Since this is the major functionality of connection manager, I'd think about having the initializer be a field of connection manager and pass it in when calling New for connection manager.